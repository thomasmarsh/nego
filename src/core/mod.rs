pub mod bitboard;
pub mod coord;
pub mod error;
pub mod game;
pub mod r#move;
pub mod move_tab;
pub mod orientation;
pub mod pieces;
pub mod ray;
pub mod square;
pub mod zobrist;
