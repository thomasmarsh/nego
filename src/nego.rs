mod ai;
mod core;

#[macro_use]
extern crate log;

fn main() -> rustyline::Result<()> {
    pretty_env_logger::init();

    info!("initializing ray LUT");
    core::ray::Rays::build_lut();

    // cli()

    //dem();
    ai::demo_minimax();
    Ok(())
}
