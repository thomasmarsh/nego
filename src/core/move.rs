use crate::core::{
    bitboard::BitBoard,
    coord::{X, Y},
    error::Error,
    move_tab::LUTEntry,
    orientation::Orientation,
    pieces::{PieceId, PieceTypeId},
    square::Square,
};

use serde::Serialize;
use std::fmt;
use std::str::FromStr;

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize)]
pub struct Move(u16);

impl Move {
    #[inline]
    pub fn new(piece: PieceId, entry: LUTEntry) -> Move {
        let p = (piece as u16) & 0b1111;
        let e = (entry.0 as u16) << 4;
        Move(e | p)
    }

    #[inline]
    pub fn get_raw_value(self) -> u16 {
        self.0
    }

    #[inline]
    pub fn get_piece(self) -> PieceId {
        PieceId::from_index(self.0 & 0b1111).unwrap()
    }

    pub fn get_lut_entry(self) -> LUTEntry {
        LUTEntry(self.0.wrapping_shr(4) as usize)
    }

    #[inline]
    pub fn position(&self) -> Square {
        self.get_lut_entry().position()
    }

    #[inline]
    pub fn orientation(&self) -> Orientation {
        self.get_lut_entry().orientation()
    }

    #[inline]
    pub fn mask(&self) -> BitBoard {
        self.get_lut_entry().mask()
    }

    // First set bit (self.position may be empty on kunoji types)
    #[inline]
    pub fn to_square(self) -> Square {
        self.mask().to_square()
    }

    #[inline]
    pub fn gaze(&self) -> BitBoard {
        self.get_lut_entry().gaze()
    }

    pub fn notation(&self) -> String {
        format!(
            "{}:{}{:?}",
            self.get_piece().piece_type_id().notation(),
            self.position().to_string().to_uppercase(),
            self.orientation()
        )
        .to_string()
    }

    pub fn parse(s: &str) -> Result<Move, Error> {
        if s.len() != 8 {
            return Err(Error::InvalidFormat);
        }
        if s.as_bytes()[4] != b':' {
            return Err(Error::InvalidFormat);
        }

        fn parse_piece_type(piece_str: &str) -> Result<PieceTypeId, Error> {
            match piece_str {
                "BOS" => Ok(PieceTypeId::Boss),
                "MAM" => Ok(PieceTypeId::Mame),
                "NOB" => Ok(PieceTypeId::Nobi),
                "KB1" => Ok(PieceTypeId::Koubaku1),
                "KB2" => Ok(PieceTypeId::Koubaku2),
                "KB3" => Ok(PieceTypeId::Koubaku3),
                "KJ1" => Ok(PieceTypeId::Kunoji1),
                "KJ2" => Ok(PieceTypeId::Kunoji2),
                "KJ3" => Ok(PieceTypeId::Kunoji3),
                "KJ4" => Ok(PieceTypeId::Kunoji4),
                _ => Err(Error::InvalidPiece),
            }
        }

        let piece_type = parse_piece_type(&s[0..4])?;
        let x = X::from_str(&s.as_bytes()[5].to_string())?;
        let y = Y::from_str(&s.as_bytes()[6].to_string())?;
        let position = Square::make_square(x, y);
        let orientation = Orientation::from_str(&s.as_bytes()[7].to_string())?;

        let piece = match piece_type {
            PieceTypeId::Boss => PieceId::Boss,
            PieceTypeId::Mame => PieceId::Mame,
            PieceTypeId::Nobi => PieceId::Nobi,
            PieceTypeId::Koubaku1 => PieceId::Koubaku1,
            PieceTypeId::Koubaku2 => PieceId::Koubaku2,
            PieceTypeId::Koubaku3 => PieceId::Koubaku3a, // may need reindexing
            PieceTypeId::Kunoji1 => PieceId::Kunoji1a,   // may need reindexing
            PieceTypeId::Kunoji2 => PieceId::Kunoji2,
            PieceTypeId::Kunoji3 => PieceId::Kunoji3,
            PieceTypeId::Kunoji4 => PieceId::Kunoji4,
        };

        let entry =
            LUTEntry::lookup(piece_type, position, orientation).ok_or(Error::LUTEntryNotFound)?;
        Ok(Move::new(piece, entry))
    }
}

impl fmt::Display for Move {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.notation())
    }
}

impl FromStr for Move {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Move::parse(s)
    }
}

pub trait MoveVisitor {
    fn visit(&mut self, m: Move);

    #[inline]
    fn bailout(&self) -> bool {
        false
    }
}

pub struct HasMoves(pub bool);

impl MoveVisitor for HasMoves {
    #[inline]
    fn visit(&mut self, _: Move) {
        self.0 = true;
    }

    #[inline]
    fn bailout(&self) -> bool {
        self.0
    }
}

pub struct MoveCounter(pub usize);

impl MoveVisitor for MoveCounter {
    #[inline]
    fn visit(&mut self, _: Move) {
        self.0 += 1;
    }
}

pub struct MoveAccumulator(pub Vec<Move>);

impl MoveAccumulator {
    #[inline]
    pub fn new() -> Self {
        Self(Vec::with_capacity(1720 >> 1))
    }
}

impl MoveVisitor for MoveAccumulator {
    #[inline]
    fn visit(&mut self, m: Move) {
        self.0.push(m);
    }
}
