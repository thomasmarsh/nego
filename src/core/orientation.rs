use crate::core::{error::Error, square::Square};

use std::str::FromStr;

#[derive(PartialEq, Clone, Copy, Debug, Eq)]
pub enum Orientation {
    S, // S, 0 (reference orientation of pieces)
    W, // E, 90
    N, // N, 180
    E, // W, 270
}

pub const NUM_ORIENTATIONS: usize = 4;

pub const ALL_ORIENTATIONS: [Orientation; 4] = [
    Orientation::S,
    Orientation::W,
    Orientation::N,
    Orientation::E,
];

impl Orientation {
    #[inline]
    pub fn at_limit(&self, square: Square) -> bool {
        use Orientation::*;
        match self {
            S => square.is_bottom_edge(),
            W => square.is_left_edge(),
            N => square.is_top_edge(),
            E => square.is_right_edge(),
        }
    }

    #[inline]
    pub fn opposite(&self) -> Orientation {
        use Orientation::*;
        match self {
            S => N,
            W => E,
            N => S,
            E => W,
        }
    }

    // if "facing" one direction, what direction is to the right?
    #[inline]
    pub fn right(&self) -> Orientation {
        use Orientation::*;
        match self {
            S => W,
            W => N,
            N => E,
            E => S,
        }
    }

    // if "facing" one direction, what direction is to the left?
    #[inline]
    pub fn left(&self) -> Orientation {
        use Orientation::*;
        match self {
            S => E,
            W => S,
            N => W,
            E => N,
        }
    }

    #[inline]
    pub fn from_index(i: u8) -> Orientation {
        match i {
            0 => Orientation::S,
            1 => Orientation::W,
            2 => Orientation::N,
            3 => Orientation::E,
            _ => unreachable!(),
        }
    }
}

impl FromStr for Orientation {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.is_empty() {
            return Err(Error::InvalidRank);
        }
        match s.chars().next().unwrap() {
            'S' => Ok(Orientation::S),
            'W' => Ok(Orientation::W),
            'N' => Ok(Orientation::N),
            'E' => Ok(Orientation::E),
            _ => Err(Error::InvalidOrientation),
        }
    }
}
