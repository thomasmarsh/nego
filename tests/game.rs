use nego::core::{
    game::State, move_tab::LUTEntry, orientation::Orientation, pieces::PieceId, r#move::Move,
    square::Square,
};

fn mk_move(piece: PieceId, pos: (u8, u8), orientation: Orientation) -> Move {
    let entry = LUTEntry::lookup(
        piece.piece_type_id(),
        Square::from_indices(pos.0 as usize, pos.1 as usize),
        orientation,
    )
    .unwrap();
    Move::new(piece, entry)
}

fn play_moves(state: &mut State, moves: &[(PieceId, (u8, u8), Orientation)]) {
    moves.iter().for_each(|(a, b, c)| {
        state.place(mk_move(*a, *b, *c));
        state.current = state.current.next();
    });
}

#[test]
fn no_capture_boss() {
    use Orientation::*;
    use PieceId::*;

    let mut state = State::new();
    let first_boss = mk_move(Boss, (0, 0), S);
    state.place(first_boss);
    state.current = state.current.next();
    state.dump();
    let occupied = state.board.black.occupied;

    play_moves(
        &mut state,
        &[
            (Boss, (0, 2), S), // W
            (Mame, (5, 5), S), // B
        ],
    );

    println!("before placement:");
    state.dump();
    state.place(mk_move(Nobi, (2, 0), W)); // W
    println!("after placement:");
    state.dump();

    assert!(state.board.black.occupied & occupied == occupied);
    assert_eq!(state.board.black.move_list[0].get_piece(), Boss);
}

#[test]
fn snapshot() {
    use rand::{seq::SliceRandom, SeedableRng};
    let mut rng = rand::rngs::StdRng::seed_from_u64(123);

    for _ in 0..20 {
        let mut state = State::new();
        let mut moves = Vec::new();
        let mut history = Vec::new();
        loop {
            history.push(state.clone());
            moves.truncate(0);
            state.get_moves(&mut moves);
            if moves.is_empty() {
                break;
            }
            let m = *moves.choose(&mut rng).unwrap();
            state.apply(m);
        }

        insta::assert_debug_snapshot!(history);
    }
}
