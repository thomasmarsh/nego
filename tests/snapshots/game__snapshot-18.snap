---
source: tests/game.rs
expression: history
---
[
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4095,
                ),
                move_list: [],
                occupied: BitBoard(
                    0,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4095,
                ),
                move_list: [],
                occupied: BitBoard(
                    0,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                0,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 0,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                ],
                occupied: BitBoard(
                    847723465015296,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4095,
                ),
                move_list: [],
                occupied: BitBoard(
                    0,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723465015296,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 18223537927370505983,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                ],
                occupied: BitBoard(
                    847723465015296,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                ],
                occupied: BitBoard(
                    202113024,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 2574906838960643479,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4062,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                ],
                occupied: BitBoard(
                    847723465015553,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                ],
                occupied: BitBoard(
                    202113024,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 12185048888884217041,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4062,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                ],
                occupied: BitBoard(
                    847723465015553,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4078,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                ],
                occupied: BitBoard(
                    1729382257112383488,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 2068451689321275075,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4058,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                ],
                occupied: BitBoard(
                    1111606255681793,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4078,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                ],
                occupied: BitBoard(
                    1729382257112383488,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 14079133647892602618,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4058,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                ],
                occupied: BitBoard(
                    1111606255681793,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4070,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                ],
                occupied: BitBoard(
                    1742893055994494976,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 11067916579541224985,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3994,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                ],
                occupied: BitBoard(
                    1111606255681841,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4070,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                ],
                occupied: BitBoard(
                    1742893055994494976,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 14780033912554836451,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3994,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                ],
                occupied: BitBoard(
                    1111606255681841,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4068,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                ],
                occupied: BitBoard(
                    1760907454503976960,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 17674681602767940304,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3986,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                    Move(
                        9251,
                    ),
                ],
                occupied: BitBoard(
                    1111606794649905,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4068,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                ],
                occupied: BitBoard(
                    1760907454503976960,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 4973095491479789294,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3986,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                    Move(
                        9251,
                    ),
                ],
                occupied: BitBoard(
                    1111606794649905,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4036,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                    Move(
                        14421,
                    ),
                ],
                occupied: BitBoard(
                    1760907454504108544,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 14927581880716847116,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2962,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                    Move(
                        9251,
                    ),
                    Move(
                        23210,
                    ),
                ],
                occupied: BitBoard(
                    1111610020069681,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4036,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                    Move(
                        14421,
                    ),
                ],
                occupied: BitBoard(
                    1760907454504108544,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 9250638666471944529,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2962,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                    Move(
                        9251,
                    ),
                    Move(
                        23210,
                    ),
                ],
                occupied: BitBoard(
                    1111610020069681,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1988,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                    Move(
                        14421,
                    ),
                    Move(
                        25163,
                    ),
                ],
                occupied: BitBoard(
                    2052515530376348160,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 3565811722731070416,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    914,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                    Move(
                        9251,
                    ),
                    Move(
                        23210,
                    ),
                    Move(
                        27035,
                    ),
                ],
                occupied: BitBoard(
                    1111610070466865,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1988,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                    Move(
                        14421,
                    ),
                    Move(
                        25163,
                    ),
                ],
                occupied: BitBoard(
                    2052515530376348160,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 3228379916114434983,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    914,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                    Move(
                        9251,
                    ),
                    Move(
                        23210,
                    ),
                    Move(
                        27035,
                    ),
                ],
                occupied: BitBoard(
                    1111610070466865,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1924,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                    Move(
                        14421,
                    ),
                    Move(
                        25163,
                    ),
                    Move(
                        13814,
                    ),
                ],
                occupied: BitBoard(
                    2052515530376397312,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 11724331851824179430,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    912,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                    Move(
                        9251,
                    ),
                    Move(
                        23210,
                    ),
                    Move(
                        27035,
                    ),
                    Move(
                        1825,
                    ),
                ],
                occupied: BitBoard(
                    1111610070470961,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1924,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                    Move(
                        14421,
                    ),
                    Move(
                        25163,
                    ),
                    Move(
                        13814,
                    ),
                ],
                occupied: BitBoard(
                    2052515530376397312,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 1952845940186488714,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    912,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                    Move(
                        9251,
                    ),
                    Move(
                        23210,
                    ),
                    Move(
                        27035,
                    ),
                    Move(
                        1825,
                    ),
                ],
                occupied: BitBoard(
                    1111610070470961,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1796,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                    Move(
                        14421,
                    ),
                    Move(
                        25163,
                    ),
                    Move(
                        13814,
                    ),
                    Move(
                        17655,
                    ),
                ],
                occupied: BitBoard(
                    2052528741695799808,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 5397454085930287579,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    400,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                    Move(
                        9251,
                    ),
                    Move(
                        23210,
                    ),
                    Move(
                        27035,
                    ),
                    Move(
                        1825,
                    ),
                    Move(
                        21001,
                    ),
                ],
                occupied: BitBoard(
                    1111816497336625,
                ),
                owned: BitBoard(
                    264917615640576,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1796,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                    Move(
                        14421,
                    ),
                    Move(
                        25163,
                    ),
                    Move(
                        13814,
                    ),
                    Move(
                        17655,
                    ),
                ],
                occupied: BitBoard(
                    2052528741695799808,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 18302307117414639530,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    400,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                    Move(
                        9251,
                    ),
                    Move(
                        23210,
                    ),
                    Move(
                        27035,
                    ),
                    Move(
                        1825,
                    ),
                    Move(
                        21001,
                    ),
                ],
                occupied: BitBoard(
                    1111816497336625,
                ),
                owned: BitBoard(
                    264917615640576,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    772,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                    Move(
                        14421,
                    ),
                    Move(
                        25163,
                    ),
                    Move(
                        13814,
                    ),
                    Move(
                        17655,
                    ),
                    Move(
                        22874,
                    ),
                ],
                occupied: BitBoard(
                    15923615593996927488,
                ),
                owned: BitBoard(
                    18229458603210572288,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 7013817543796471933,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    272,
                ),
                move_list: [
                    Move(
                        544,
                    ),
                    Move(
                        12869,
                    ),
                    Move(
                        5858,
                    ),
                    Move(
                        15238,
                    ),
                    Move(
                        9251,
                    ),
                    Move(
                        23210,
                    ),
                    Move(
                        27035,
                    ),
                    Move(
                        1825,
                    ),
                    Move(
                        21001,
                    ),
                    Move(
                        15975,
                    ),
                ],
                occupied: BitBoard(
                    1111816497338685,
                ),
                owned: BitBoard(
                    264917615640576,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    772,
                ),
                move_list: [
                    Move(
                        256,
                    ),
                    Move(
                        12020,
                    ),
                    Move(
                        7331,
                    ),
                    Move(
                        3297,
                    ),
                    Move(
                        14421,
                    ),
                    Move(
                        25163,
                    ),
                    Move(
                        13814,
                    ),
                    Move(
                        17655,
                    ),
                    Move(
                        22874,
                    ),
                ],
                occupied: BitBoard(
                    15923615593996927488,
                ),
                owned: BitBoard(
                    18229458603210572288,
                ),
            },
            boss: BitBoard(
                847723667128320,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 17644215520126936561,
        capture_flag: false,
    },
]
